h/traversal/group_membership.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+
+from pyramid.httpexceptions import HTTPNotFound
+
+from h.exceptions import InvalidUserId
+from h.models import Group, GroupMembership, User
+
+
+@dataclass
+class GroupMembershipContext:
+    group: Group
+    user: User
+    membership: GroupMembership | None
+
+
+def group_membership_api_factory(request) -> GroupMembershipContext:
+    user_service = request.find_service(name="user")
+    group_service = request.find_service(name="group")
+    group_members_service = request.find_service(name="group_members")
+
+    userid = request.matchdict["userid"]
+    pubid = request.matchdict["pubid"]
+
+    def get_user() -> User | None:
+        if userid == "me":
+            if request.authenticated_userid:
+                return user_service.fetch(request.authenticated_userid)
+
+            return None
+
+        try:
+            return user_service.fetch(userid)
+        except InvalidUserId:
+            return None
+
+    user = get_user()
+
+    if not user:
+        raise HTTPNotFound(f"User not found: {userid}")
+
+    group = group_service.fetch(pubid)
+
+    if not group:
+        raise HTTPNotFound(f"Group not found: {pubid}")
+
+    membership = group_members_service.get(group, user)
+
+    if not membership and request.method != "POST":
+        raise HTTPNotFound(f"Membership not found: ({pubid}, {userid})")
+
+    return GroupMembershipContext(group=group, user=user, membership=membership)
```

Comment:
The `factory="h.traversal.group_membership_api_factory"` in the route config means that Pyramid will call `group_membership_api_factory()` to get the `context` object for requests that match the `"api.group_member"` route.

This function either returns a `GroupMembershipContext` object or raises `HTTPNotFound` which causes Pyramid to 404 the request without ever calling the security policy or view.

This has some big advantages:

1. There's a nice separation of concerns where this factory is responsible for finding the `group`, `user` and `membership` (and handling what to do when they don't exist) and then the security policy and view can assume that if they're called then the the `group`, `user` and `membership` _do_ exist.
2. This factory can do the DB queries to look up these objects once and other code can access them via the `context` argument, avoiding duplicate DB queries.
3. h can return 404s for resources that don't exist, and 401s for resources that the request doesn't have permission to act on.

If you don't use traversal you get a problem. If someone makes an API request with an unknown group ID or unknown user ID in the URL, or with a group ID and user ID for which there is no membership, then Pyramid will end up calling the security policy which must return either `Allowed` or `Denied`, neither of which is a good choice:

1. It doesn't seem quite right to me for the security policy to return `Denied` in this situation. It's not that the request doesn't have permission to edit the membership, the problem is that the membership doesn't exist. Returning `Denied` would cause Pyramid to send a 401 response, when a 404 would be more appropriate.

2. The security policy could return `Allowed` and then the view could raise `HTTPNotFound`. This would return the correct response (404) but it seems wrong to me for the security policy to say that the request is permitted to edit a membership that doesn't exist and I imagine there could even be race conditions resulting in security issues (if a membership with the matching IDs gets created by a concurrent request and it happens to be a membership that the API request should *not* have been permitted to edit).